docs: Update contributing and installation guides (#3656)
narwhals-dev/narwhals: CONTRIBUTING.md

If you've got experience with open source contributions, the following instructions might suffice:

- Install [uv](https://docs.astral.sh/uv/getting-started/installation/) (required).
- Clone repo: `git clone git@github.com:narwhals-dev/narwhals.git narwhals-dev`
- `cd narwhals-dev/`
- `git remote rename origin upstream`
- `git remote add origin <your fork goes here>`
- `uv sync --group local-dev` (creates `.venv` and installs project + dev deps)
- Install pre-commit as a git hook: `uv run pre-commit install`
- To run tests: `uv run pytest`
- To run all linting checks: `uv run pre-commit run --all-files`
- To run static typing checks: `make typing`

For more detailed and beginner-friendly instructions, see below!

## Prerequisites

> Narwhals uses [uv](https://docs.astral.sh/uv/) as its package and environment manager. You must have `uv` installed to following the rest of this guide. The Makefile targets, dependency groups, and lockfile (`uv.lock`) all assume `uv` is present.

Reviewer: ```suggestion
Narwhals uses [uv](https://docs.astral.sh/uv/) as its package and environment manager. You must have `uv` installed to follow the rest of this guide. The Makefile targets, dependency groups, and lockfile (`uv.lock`) all assume `uv` is present.
```

To install `uv`, follow the official instructions at [uv installation](https://docs.astral.sh/uv/getting-started/installation/). If you already have `uv`, keep it up-to-date with:

```terminal
uv self update
```

### Optional: Java for PySpark tests

> If you want to run PySpark-related tests, you'll also need to have Java installed. Refer to the [Spark documentation](https://spark.apache.org/docs/latest/#downloading) for more information.

Reviewer 2: I yearn for a far off future where we have a command for this 😂

Reviewer: `pixi` or `conda` [can install the correct version of Java](https://anaconda.org/channels/conda-forge/packages/openjdk/overview), it's how I usually end up installing for Spark when I need to test something with it locally.

Reviewer 2: @camriddell this definitely warrants it's own issue/PR - but I'm curious about using `pixi` for it

### 5. Setting up your environment

With `uv` already installed (see [Prerequisites](#prerequisites)), set up the project. `uv sync` will create a `.venv` and install the project together with the `local-dev` dependency group (fast-ish core libraries and dev dependencies):

```terminal

> uv sync --group local-dev

Reviewer 2: I'm kinda confused

> `uv sync` will create a `.venv` and install the project together with the `local-dev`

But what follows is `uv sync --group local-dev`.

I think the way you've described it sounds like [default-groups](https://docs.astral.sh/uv/concepts/projects/dependencies/#default-groups).

If we had this instead, then the only people who need to pick a `--group` are those already picky about them anyway 😉 

```toml
[tool.uv]
default-groups = ["local-dev"]
```

Author: Mostly bad wording from myself, adjusting now

```

If you also want to test other libraries like Dask and PySpark, add their extras:

```terminal
uv sync --group local-dev --extra dask --extra pyspark
```

You don't need to pin a Python version manually: `uv` will pick a compatible interpreter (downloading one if needed) based on `pyproject.toml`. If you want a specific version, pass `--python 3.12` (or similar) to `uv sync`.

You then have two options to run commands inside the project environment:

- Activate the venv (`. .venv/bin/activate` on Linux/macOS, `.\.venv\Scripts\activate` on Windows) and run commands directly (e.g. `pytest`).

> - Prefix commands with `uv run` (e.g. `uv run pytest`). This is the recommended approach as `uv` will re-sync the environment if needed.

Reviewer 2: 1. Recommended should go first (more important)
2. Maybe just leave out the environment activation part instead? (less important)

We can't currently test in CI against cuDF, but you can test it manually in Kaggle using GPUs. Please follow this [Kaggle notebook](https://www.kaggle.com/code/marcogorelli/testing-cudf-in-narwhals) to run the tests.

### Dependency management with uv

A few useful `uv` workflows when contributing:

- Add a dependency to a group (e.g. `local-dev`): `uv add --group local-dev <package>`
- Add a dependency to an extra (e.g. `dask`): `uv add --optional dask <package>`
- Remove a dependency: `uv remove <package>` (with `--group`/`--optional` as appropriate)
- Refresh the lockfile without changing pins: `uv lock`

> - Upgrade everything: `uv lock --upgrade`

Reviewer 3: thansk for including these 🙌

Reviewer 2: @MarcoGorelli re (https://github.com/narwhals-dev/narwhals/pull/3608#issuecomment-4412835603)

It's all probably just gonna be *unlearning* `pip`

> - Upgrade everything: `uv lock --upgrade`

Reviewer 2: One part you might want to add is that the order is:

```terminal
uv lock
uv sync
```

I'm dumb and have forgotten this like 12 times 😭 

Also some links:
- https://docs.astral.sh/uv/concepts/projects/sync/#syncing-the-environment
- https://docs.astral.sh/uv/concepts/projects/sync/#syncing-development-dependencies
- https://docs.astral.sh/uv/concepts/projects/sync/#upgrading-locked-package-versions
- https://docs.astral.sh/uv/reference/cli/#uv-sync
- https://docs.astral.sh/uv/reference/cli/#uv-lock

All of these update `pyproject.toml` and `uv.lock`. Commit both files together when changing dependencies.

### Static typing

> We run `mypy`, `pyright`, and `pyrefly` in CI. All of them are included in the `typing` dependency group and are installed on-demand when running `make typing`.

Reviewer 2: ```suggestion
We run `mypy`, `pyright`, and `pyrefly` in CI. All of them are included in `--group typing` and are installed as needed when running `make typing`.
```

Run them with:

```console
make typing
```

> to verify type completeness / correctness. Under the hood this runs `uv run --group typing pyright`, `uv run --group typing mypy`, and `uv run --group typing pyrefly check`.

Reviewer 2: I think the description of the commands can be left out?

### 9. Building the docs

To serve the docs locally, run:

```terminal
make docs-serve
```

and then open the link provided in a browser.

The docs should refresh when you make changes. If they don't, press `ctrl+C`, and then run:

```terminal
uv run --group docs zensical build --clean

> make docs-serve

Reviewer 2: Could this be a single command?

I'm actually a little confused that this doesn't build 🤔 

https://github.com/narwhals-dev/narwhals/blob/1d75ff05c814ea316e36d01673d27badb29c8e2b/Makefile#L22-L26

Reviewer 2: Recently I discovered @koaning posted [this video](https://www.youtube.com/watch?v=n-L2p-poGUk) that showed some commands I added in (https://github.com/vega/altair/pull/3723) 😅 

So there's some ideas for other commands too

### 9. Building the docs

To serve the docs locally, run:

```terminal
make docs-serve
```

and then open the link provided in a browser.

The docs should refresh when you make changes. If they don't, press `ctrl+C`, and then run:

```terminal

> uv run --group docs zensical build --clean && make docs-serve

Reviewer 2: https://github.com/narwhals-dev/narwhals/pull/3656#discussion_r3310236048
> Could this be a single command?

Sorry for not being clear
I meant could this be another command **in the Makefile?**

So this section would read like:

>
> ### 9. Building the docs
> 
> To serve the docs locally, run:
> 
> ```terminal
> make docs-serve
> ```
> 
> and then open the link provided in a browser.
> 
> The docs should refresh when you make changes. If they don't, press `ctrl+C`, and then run:
> 
> ```terminal
> make docs-<other-name>
> ```

Author: The hard part is to come up with a good "\<other-name\>" part 😅

Reviewer 2: 🎁
```terminal
make docs-clean
make docs-build
make docs-rebuild
make docs-plz
make docs-i-mean-it-this-time
```

Author: I went for `docs-clean-serve` because it mirrors the existing `docs-serve` and makes the "clean + serve" semantics obvious in `make help`. Others are shorter but lose the hint that it also serves the docs locally.
